_posts/2025-06-03-why-open-source-maintainers-thrive-in-the-llm-era.md

```diff
@@ -0,0 +1,125 @@
+With LLM agents, you can even instruct them to self-verify their output by executing relevant commands or tests.
+For example, when working on Homebrew I'll ask agents to run `brew style` (for RuboCop code linting), `brew typecheck` (for static type checking) and `brew tests` (for unit tests) to automatically verify behaviour.
+
+## 📵 LLM Modes
+
+Based on these learnings, I've developed a few distinct "LLM modes" in my workflow:
+
+1. Constantly: I'm writing code, my editor of choice (Cursor, today) is just providing a nice autocomplete and quick, inline lookup for stuff I've forgotten.
+2. Regularly: I'm blocked and would normally be Googling or looking at StackOverflow for a solution.
+   Instead, I'll now ask the LLM in my editor or ChatGPT for a solution or ideas.
+   This excels at rapidly deciphering confusing error messages or sifting through large log dumps to pinpoint the source of issues.
+3. Rarely: I get the LLM to generate a large amount of boring and fairly trivial code based on a lot of initial research I've done.
+   An example would be in the [MCP Server for Homebrew](https://github.com/Homebrew/brew/pull/20041) where I got Cursor to generate a first pass for a couple of methods.
+   I then edited this extensively, made it look and work how I wanted through manual edits and LLM refactoring.
+   After this, I used it to similarly generate most of the first versions of the unit tests to hit decent code coverage.
+   Throughout this process, I maintain frequent `git` commits, enabling me to use `git diff` for careful review of each generated change.
+4. Rarest: I haven't decided how I want something to work yet so I get LLM to generate all of the code, don't even look at the code and repeatedly change prompts based on the generated UI or CLI output.
+   Once the functionality is achieved, I then step away, and the following day, conduct a thorough line-by-line review and editing pass.
+   This was similar to the writing workflow I took when writing [Git in Practice](https://mikemcquaid.com/gitinpractice/) where I'd write until I hit a page count without reading things back and then do an edit pass the next day.
+
+I would say I do 1) hourly, 2) daily, 3) weekly and 4) monthly at this point.
+This pattern emerged from a continuous assessment of what maximises my personal and team productivity, balancing rapid development with preventing high-impact bugs for users.
+
+## 🪩 Reflection
+
+I've found LLM tooling like Cursor and ChatGPT to be an essential part of my workflow.
+Depending on the task, I'd say they provide anywhere between a 1% to 100% speedup.
+For me, ChatGPT has replaced 99% of my prior Google usage.
+
+I'm genuinely curious to see what the future holds for LLMs.
+I suspect with AI hype we're seeing a certain amount of the
+[Gell-Mann amnesia effect](https://en.wikipedia.org/wiki/Gell-Mann_amnesia_effect)
+where people are bullish about AI disruption on the areas they have the least context on.
```

Comment:
I enjoyed reading about Gell-Mann amnesia effect as I'm a Crichton fan. But I think that most readers are lazy and probably won't read the link and therefore the preceding sentence don't really make much sense. In fact, even after reading it and understanding Gell-Mann amnesia effect, I feel like I still don't understand this sentence that well 😆.